### PR TITLE
Array indexing

### PR DESCRIPTION
**Pull request by endremborza**
added a test for dask array loc
added a function and a test for the indexer whether it is an instance of dask array

solves part of #4859

### CODE REVIEW

**Comment by mrocklin** [comment]
In principle this seems fine to me.  I'm curious though, what happens if the user tries to index with a dask array that is not well aligned with the dask dataframe?  For example 

1.  They might generate a dask array with a different number of partitions
2.  Or with a different number of rows per partition
3.  Or they might pass in an array with integer dtype and try to index by enumerating rows

In these cases (or at least 1,3, which are easily detectable) we probably want to continue raising somewhat informative error messages.

**Comment by endremborza** [comment]
Well, in the current behavior, the dask array is assumed to be a scalar through quite a few function calls, so i don't think the error messages would be worse this way.

I will look into writing some tests for these cases, I think 3 is caught well enough with the boolean meta.

However, continued work on this raises more general questions of dask array -> series conversion, which as I see it is not implemented very thoroughly.

**Comment by TomAugspurger** [comment]
> raises more general questions of dask array -> series conversion

I'm fine with just supporting cases where the number of blocks / partitions match between the indexer and the target series. We can raise when they don't match.

File: dask/dataframe/indexing.py [hunk 1]
```diff
@@ -133,6 +136,16 @@ def _loc_series(self, iindexer, cindexer):
...
                                        token='loc-series', meta=meta)
 
+    def _loc_array(self, iindexer, cindexer):
+        meta = self._make_meta(iindexer, cindexer)
+        iindexer_series = Series(iindexer.dask,
+                                 iindexer.name,
+                                 pd.Series([], dtype=bool),
+                                 self.obj.divisions)
+
+        return self.obj.map_partitions(methods.loc, iindexer_series, cindexer,
```
[the remark this answers is not in the document]
Comment (endremborza) [reply dask/dataframe/indexing.py]: if i just change-it, it breaks the test, because of the meta

**Comment by endremborza** [comment]
>     1. They might generate a dask array with a different number of partitions

this gives ValueError: Shapes do not align, which I find much more informative than any previous one, added a test for it.

3, raises a KeyError, which is not very informative, suggestions are welcome

2, takes more time and integration of dataframe and array tests, i think

File: dask/dataframe/indexing.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TomAugspurger) [reply dask/dataframe/indexing.py]: Can you expand on this?

[the remark this answers is not in the document]
Comment (endremborza) [reply dask/dataframe/indexing.py]: nevermind, I changed something else that broke the test and attributed it to the meta. I did it with calling ```self._loc_series``` now

**Comment by mrocklin** [comment]
I gave this a shot and ran into some issues.  They're recroded below:

```python
In [1]: import dask

In [2]: df = dask.datasets.timeseries()

In [3]: df.loc[df.x.values > 0]
Out[3]:
Dask DataFrame Structure:
                   id    name        x        y
npartitions=30
2000-01-01      int64  object  float64  float64
2000-01-02        ...     ...      ...      ...
...               ...     ...      ...      ...
2000-01-30        ...     ...      ...      ...
2000-01-31        ...     ...      ...      ...
Dask Name: loc-series, 150 tasks

In [4]: df.loc[df.x.values.blocks[:5] > 0]  # surprisingly, this works, even though the indexer has fewer partitions
Out[4]:
Dask DataFrame Structure:
                   id    name        x        y
npartitions=30
2000-01-01      int64  object  float64  float64
2000-01-02        ...     ...      ...      ...
...               ...     ...      ...      ...
2000-01-30        ...     ...      ...      ...
2000-01-31        ...     ...      ...      ...
Dask Name: loc-series, 130 tasks

In [5]: df.loc[df.x.values.blocks[:5] > 0].compute()
```

```python-traceback
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
pandas/_libs/index.pyx in pandas._libs.index.DatetimeEngine.get_loc()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.Int64HashTable.get_item()

TypeError: an integer is required

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
~/miniconda/envs/dev/lib/python3.7/site-packages/pandas/core/indexes/base.py in get_loc(self, key, method, tolerance)
   2601             try:
-> 2602                 return self._engine.get_loc(key)
   2603             except KeyError:

pandas/_libs/index.pyx in pandas._libs.index.DatetimeEngine.get_loc()

pandas/_libs/index.pyx in pandas._libs.index.DatetimeEngine.get_loc()

pandas/_libs/index.pyx in pandas._libs.index.DatetimeEngine._date_check_type()

KeyError: 'gt-0f4c9ac82163c0ea93b9d01edd0f1dbd'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
pandas/_libs/index.pyx in pandas._libs.index.DatetimeEngine.get_loc()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.Int64HashTable.get_item()

TypeError: an integer is required
```

Also, should we also support this in `__getitem__` calls?

```python
In [6]: df[df.x.values > 0]
```
```python-traceback
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-6-87dd5b65cde0> in <module>
----> 1 df[df.x.values > 0]

~/workspace/dask/dask/dataframe/core.py in __getitem__(self, key)
   2693             graph = HighLevelGraph.from_collections(name, dsk, dependencies=[self, key])
   2694             return new_dd_object(graph, name, self, self.divisions)
-> 2695         raise NotImplementedError(key)
   2696
   2697     def __setitem__(self, key, value):

NotImplementedError: dask.array<gt, shape=(nan,), dtype=bool, chunksize=(nan,)>
```

It looks like we also fail to check for matching divisions in the series case (though this likely predates this PR)

```python
In [7]: df[df.x.partitions[:5] > 0]
Out[7]:
Dask DataFrame Structure:
                   id    name        x        y
npartitions=30
2000-01-01      int64  object  float64  float64
2000-01-02        ...     ...      ...      ...
...               ...     ...      ...      ...
2000-01-30        ...     ...      ...      ...
2000-01-31        ...     ...      ...      ...
Dask Name: getitem, 220 tasks

In [8]: df[df.x.partitions[:5] > 0].compute()
```

```python-traceback
/Users/mrocklin/workspace/dask/dask/core.py:119: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  return func(*args2)
---------------------------------------------------------------------------
IndexingError                             Traceback (most recent call last)
<ipython-input-8-f5c733dd3417> in <module>
----> 1 df[df.x.partitions[:5] > 0].compute()

~/workspace/dask/dask/base.py in compute(self, **kwargs)
    154         dask.base.compute
    155         """
--> 156         (result,) = compute(self, traverse=False, **kwargs)
    157         return result
    158

~/workspace/dask/dask/base.py in compute(*args, **kwargs)
    396     keys = [x.__dask_keys__() for x in collections]
    397     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 398     results = schedule(dsk, keys, **kwargs)
    399     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    400

~/workspace/dask/dask/threaded.py in get(dsk, result, cache, num_workers, pool, **kwargs)
     74     results = get_async(pool.apply_async, len(pool._pool), dsk, result,
     75                         cache=cache, get_id=_thread_get_id,
---> 76                         pack_exception=pack_exception, **kwargs)
     77
     78     # Cleanup pools associated to dead threads

~/workspace/dask/dask/local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    460                         _execute_task(task, data)  # Re-execute locally
    461                     else:
--> 462                         raise_exception(exc, tb)
    463                 res, worker_id = loads(res_info)
    464                 state['cache'][key] = res

~/workspace/dask/dask/compatibility.py in reraise(exc, tb)
    111         if exc.__traceback__ is not tb:
    112             raise exc.with_traceback(tb)
--> 113         raise exc
    114
    115     import pickle as cPickle

~/workspace/dask/dask/local.py in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    228     try:
    229         task, data = loads(task_info)
--> 230         result = _execute_task(task, data)
    231         id = get_id()
    232         result = dumps((result, id))

~/workspace/dask/dask/core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

~/workspace/dask/dask/optimization.py in __call__(self, *args)
    940                              % (len(self.inkeys), len(args)))
    941         return core.get(self.dsk, self.outkey,
--> 942                         dict(zip(self.inkeys, args)))
    943
    944     def __reduce__(self):

~/workspace/dask/dask/core.py in get(dsk, out, cache)
    147     for key in toposort(dsk):
    148         task = dsk[key]
--> 149         result = _execute_task(task, cache)
    150         cache[key] = result
    151     result = _execute_task(out, cache)

~/workspace/dask/dask/core.py in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

~/miniconda/envs/dev/lib/python3.7/site-packages/pandas/core/frame.py in __getitem__(self, key)
   2906         # Do we have a (boolean) 1d indexer?
   2907         if com.is_bool_indexer(key):
-> 2908             return self._getitem_bool_array(key)
   2909
   2910         # We are left with two options: a single key, and a collection of keys,

~/miniconda/envs/dev/lib/python3.7/site-packages/pandas/core/frame.py in _getitem_bool_array(self, key)
   2955         # check_bool_indexer will throw exception if Series key cannot
   2956         # be reindexed to match DataFrame rows
-> 2957         key = check_bool_indexer(self.index, key)
   2958         indexer = key.nonzero()[0]
   2959         return self._take(indexer, axis=0)

~/miniconda/envs/dev/lib/python3.7/site-packages/pandas/core/indexing.py in check_bool_indexer(ax, key)
   2484         mask = isna(result._values)
   2485         if mask.any():
-> 2486             raise IndexingError('Unalignable boolean Series provided as '
   2487                                 'indexer (index of the boolean Series and of '
   2488                                 'the indexed object do not match')

IndexingError: Unalignable boolean Series provided as indexer (index of the boolean Series and of the indexed object do not match
```

File: dask/dataframe/indexing.py [hunk 2]
```diff
@@ -133,6 +136,13 @@ def _loc_series(self, iindexer, cindexer):
         return self.obj.map_partitions(methods.loc, iindexer, cindexer,
                                        token='loc-series', meta=meta)
 
+    def _loc_array(self, iindexer, cindexer):
+        iindexer_series = Series(iindexer.dask,
+                                 iindexer.name,
+                                 pd.Series([], dtype=bool),
+                                 self.obj.divisions)
```
Comment (mrocklin) [inline dask/dataframe/indexing.py]: I wonder if the `to_dask_dataframe` method might be more robust here.

Comment (endremborza) [reply dask/dataframe/indexing.py]: tried it, works.

**Comment by endremborza** [comment]
added a simple error message, a new test, and rewrote it to to_dask_dataframe.

> Also, should we also support this in ```__getitem__``` calls?

I think that would be useful, yes.

**Comment by jcrist** [comment]
Checking in here, what's the status of this? Is this ok to merge (all tests are passing)?

**Comment by martindurant** [comment]
Ping again - there was a commit since the last review, so request @mrocklin and/or @TomAugspurger have another look, and merge if done.

**Comment by mrocklin** [comment]
I've made a couple of tiny changes .  Merging in.  Thanks @endremborza !

Also, I notice that this is your first code contribution to this repository.  Welcome!

**Comment by endremborza** [comment]
thanks @mrocklin !

I quite like the project, I think I will look to contribute more if I find something interesting.

**Pull request merged**